File: src/main/java/de/codecentric/springnewspaper/service/NewspaperService.java
package de.codecentric.springnewspaper.service;

import de.codecentric.springnewspaper.domain.Newspaper;
import de.codecentric.springnewspaper.gateway.NewspaperGateway;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.NoSuchElementException;

@Component
public class NewspaperService {
    private final NewspaperGateway newspaperGateway;

    public NewspaperService(@Autowired NewspaperGateway newspaperGateway) {
        this.newspaperGateway = newspaperGateway;
    }

    public Newspaper getNewspaperById(Integer id) {
        Newspaper newspaper = newspaperGateway.get(id);

        if (newspaper.getPublicationDate().isAfter(Instant.now())) {
            throw new NoSuchElementException("Newspaper is not yet published.");
        }

        return newspaper;
    }
}

File: src/main/kotlin/de/codecentric/springnewspaper/service/NewspaperService.kt
package de.codecentric.springnewspaper.service

import de.codecentric.springnewspaper.domain.Newspaper
import de.codecentric.springnewspaper.gateway.NewspaperGateway
import java.time.Instant
import org.springframework.stereotype.Component

@Component
class NewspaperService(val gateway: NewspaperGateway) {

    fun getNewspaperById(id: Int): Newspaper {
        val newspaper = gateway.get(id)

        // Forbid access to newspapers that will be published in the future
        if (newspaper.publicationDate > Instant.now()) {
            throw NoSuchElementException("Newspaper is not yet published.")
        }

        return newspaper
    }
}
